lib/remote/actionshandler.cpp

```diff
@@ -123,7 +128,7 @@ bool ActionsHandler::HandleRequest(
 		statusCode = *okStatusCodes.begin();
 	} else if (nonOkSize == 1u) {
 		statusCode = *nonOkStatusCodes.begin();
-	} else if (okSize >= 2u && nonOkSize == 0u) {
+	} else if ((okSize == 0u || okSize >= 2u) && nonOkSize == 0u) {
```

Comment:
That check could be simplified to just `nonOkSize == 0u` as `okSize == 1u && nonOkSize == 0u` is covered by the first branch already.

But I'm not entirely sure about the 0 results case. I have a slight tendency towards returning 404 in this case as I think you usually won't perform API queries were you expect no objects to be affected, so that should probably be considered an error condition. Any second opinions on this? @Al2Klimov @lippserd

Reply:
> you usually won't perform API queries were you expect no objects to be affected

We do. We've implemented an external "watchdog" that handles late `last_check`s and nonsense `next_check`s to tackle #9278 and #9280.  In both cases, the expected match set is 0 objects. (It does a periodic `process-check-result` with filter `f"service.last_check>{now}-service.check_interval  && service.status != 3"` to set late checks to state `UNKNOWN`, and does a force `reschedule-check` with filter `f"service.next_check>{now}+service.check_interval || service.next_check<{now}"` on kernel [timerfd](https://man7.org/linux/man-pages/man2/timerfd_create.2.html) notification about time leaps.)

But I agree, the more common case is to have matches, and `404` is a obvious choice. The only issue I see with `404` is we couldn't tell "action misspelled (error)" from "0 matches (expected)" in our use case explained above.

Reply:
Mapping every possible outcome to the very limited set of HTTP status codes will always be lossy. So some cases will be non-ideal no matter what. But I think I'd also be fine with 200 as that's what gets returned for https://localhost:5665/v1/objects/hosts?filter=false right now, so it would at least be consistent with that.

Reply:
> Mapping every possible outcome to the very limited set of HTTP status codes will always be lossy

True. As a compromise, one could encode important details into the HTTP response body to keep responses at useful abstraction level.